### PR TITLE
sklearn runinference regression example

### PR DESCRIPTION
**Pull request by ryanthompson591**
This is an example of a regression model using an sklearn pipeline.

This illustrates how to separate examples based on what features are available in the example and then using an sklearn pipeline made for that data.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ryanthompson591** [comment]
R: @AnandInguva

**Comment by ryanthompson591** [comment]
R: @tvalentyn

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""A pipeline that uses RunInference API on a regression about housing prices.
+
+This example uses the japanese housing data from kaggle.
```
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:20]: Can we provide the link to the Kaggle repo ?

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+Since the data has missing fields, this example illustrates how to split
+data and assign it to appropriate models. The predictions are then recombined.
```
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:23]: different models are trained on different features of data right? may be we can mention something like that in the comments as well?
```suggestion
data and assign it to the models that are trained on different subsets of features. The predictions are then recombined.
```

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+def sort_by_features(dataframe, max_size):
+  """ Partitions the dataframe by what data it has available."""
+  for i, model in enumerate(MODELS):
+    required_features = dataframe[model['required_features']]
+    if not required_features.isnull().any().any():
+      return i
+  return -1
+
+
+def load_dataframe(file_name: str) -> Iterable[pandas.DataFrame]:
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:81]: Can we use a method instead of `DoFn` object for a `ParDo` that will achieve the same performance? just curious

From the examples I saw, `beam.ParDo()` is taking a `DoFn` object

[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:81]: Here, it says it needs `DoFn` object. So may be we will change to DoFn class

https://beam.apache.org/documentation/programming-guide/#:~:text=When%20you%20apply%20a%20ParDo%20transform%2C%20you%E2%80%99ll%20need%20to%20provide%20user%20code%20in%20the%20form%20of%20a%20DoFn%20object.%20DoFn%20is%20a%20Beam%20SDK%20class%20that%20defines%20a%20distributed%20processing%20function.

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+This example uses the japanese housing data from kaggle.
+https://www.kaggle.com/datasets/nishiodens/japan-real-estate-transaction-prices
+
+Since the data has missing fields, this example illustrates how to split
+data and assign it to the models that are trained on different subsets of
+features. The predictions are then recombined.
+
+In order to set this example up, you will need two things.
+1. Build models (or use ours) and reference those via the model directory.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:28]: I am not seeing specific instructions in this PR, and no changes to sdks/python/apache_beam/examples/inference/README.md - what is the plan?

Comment (tvalentyn) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:28]: Also, what is the plan for adding an IT for this example?

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--input',
+      dest='input',
+      required=True,
+      help='A metadata file with all models, with references to models and '
+      'information about all the files and data.')
+  parser.add_argument(
+      '--model_path',
+      dest='model_path',
```
[the remark this answers is not in the document]
Comment (ryanthompson591) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:109]: Hrm, I prefer model_path.  I think even though there are many models each model will have the same path. I was trying to think of why, so if you had a path containing libraries, you would probably call it lib_path or library_path rather than libraries_path.

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py (hunk 4 above)
Comment (AnandInguva) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:28]: We need to have a regression test for every example to catch any early bugs/regressions in the transform. 

Can you create one regression test with a smaller data and add that one to this (class)[https://github.com/apache/beam/blob/2c8e7eb7a39cbe3a1678a5c6b8b3f8700d4d8706/sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py#L41]

Comment (ryanthompson591) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:28]: I was planning to do that in a separate PR, but it's probably fine to do that here.

File: sdks/python/apache_beam/ml/inference/sklearn_inference_regression_it_test.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  with FileSystems().open(filepath) as f:
+    lines = f.readlines()
+  lines = [l.decode('utf-8').strip('\n') for l in lines]
+  return sorted(lines)
+
+
+@pytest.mark.skip
+@pytest.mark.uses_sklearn
+@pytest.mark.it_postcommit
+class SklearnInferenceRegression(unittest.TestCase):
```
Comment (ryanthompson591) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_regression_it_test.py:42]: @AnandInguva What is the most efficient way you've found to run these tests?  Normally I was just using the script to run locally, but since this test relies on data in gcs that is not public I was wondering if you've found an efficient way to run your tests.

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/sklearn_inference_regression_it_test.py:42]: You can do one thing. 

- Remove the skip decorator and run `./gradlew :sdks:python:test-suites:direct:py38:inferencePostCommitIT`. This gradle task will cover this test since it has `uses_sklearn` marker. 

Right now it skips the all the tests since the `skip` decorator is on the class. You can move this decorator on to the test methods instead.
Also, can you add this test to the already existing `sklearn_inference_it_test.py`?.  This _it_test.py is intended for all the sklearn inference regression tests.
https://github.com/apache/beam/blob/8a32df19e2c8e6b5dd5e5142b62b6ec6db7cc2bd/sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py#L41

Comment (ryanthompson591) [reply sdks/python/apache_beam/ml/inference/sklearn_inference_regression_it_test.py]: OK, I made this change, it makes sense to keep these tests together. 

I wasn't able to get hte test to run with gradlew, so I just used my local files and ran the test.  All I really needed to change was a flag to modify the numbers appended to the output file and the test passed.

**Comment by AnandInguva** [comment]
Run Python 3.7 PostCommit

**Comment by AnandInguva** [comment]
The inference tests will run on lowest and highest python versions. Python 37 and 39 as of now

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    # missing data. This example illustrates how to assign predictions to
+    # different models depending upon what data is available.
+    [all, floor_area, stations, no_features] = (
+        loaded_data
+        | 'Partition' >> beam.Partition(sort_by_features, len(MODELS)))
+
+    model_path = known_args.model_path
+    prediction_1 = all | inference_transform('all_features', model_path)
+    prediction_2 = floor_area | inference_transform('floor_area', model_path)
+    prediction_3 = stations | inference_transform('stations', model_path)
```
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:150]: This model couldn't be found in the apache-beam-ml bucket. I can see the rest three and `frontage.pickle`

File: sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py [hunk 8]
```diff
@@ -23,6 +23,7 @@
 
 import pytest
 
+from apache_beam.examples.inference import sklearn_japanese_housing_regression
 from apache_beam.examples.inference import sklearn_mnist_classification
 from apache_beam.io.filesystems import FileSystems
 from apache_beam.testing.test_pipeline import TestPipeline
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py:29]: Also, after all the imports can you add this as well?

```
# pylint: disable=wrong-import-order, wrong-import-position, ungrouped-imports, unused-import
try:
  from apache_beam.io.gcp.gcsfilesystem import GCSFileSystem
except ImportError:
  raise unittest.SkipTest('GCP dependencies are not installed')
```

Some of the tests in tox environment will run the sklearn test without the `gcp` dependencies of beam. Without `apache_beam[gcp]`, it will result in a failure. 

We have other tox envs which run sklearn tests with `apache_beam[gcp]`. That env will capture your test.

File: sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py [hunk 9]
```diff
@@ -35,10 +36,17 @@ def process_outputs(filepath):
...
+  lines = [l.decode('utf-8').strip('\n') for l in lines]
+  return sorted(lines)
+
+
 @pytest.mark.uses_sklearn
 @pytest.mark.it_postcommit
 class SklearnInference(unittest.TestCase):
+  @pytest.mark.skip
   def test_sklearn_mnist_classification(self):
     test_pipeline = TestPipeline(is_integration_test=False)
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py:51]:
```suggestion
    test_pipeline = TestPipeline(is_integration_test=True)
```

File: sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py [hunk 10]
```diff
@@ -70,6 +78,27 @@ def test_sklearn_mnist_classification(self):
       true_label, expected_prediction = expected_outputs[i].split(',')
       self.assertEqual(predictions_dict[true_label], expected_prediction)
 
+  def test_sklearn_regression(self):
+    test_pipeline = TestPipeline(is_integration_test=False)
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py:82]:
```suggestion
    test_pipeline = TestPipeline(is_integration_test=True)
```

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py (hunk 7 above)
Comment (ryanthompson591) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:150]: yeah, that was a mistake, I've added the model.

File: sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py [hunk 11]
```diff
@@ -70,6 +78,27 @@ def test_sklearn_mnist_classification(self):
...
+    }
+    sklearn_japanese_housing_regression.run(
+        test_pipeline.get_full_options_as_args(**extra_opts),
+        save_main_session=False)
+    self.assertEqual(FileSystems().exists(output_file), True)
+
+    expected_output_filepath = 'gs://apache-beam-ml/testing/expected_outputs/japanese_housing_subset.txt'  # pylint: disable=line-too-long
+    expected = file_lines_sorted(expected_output_filepath)
+    actual = file_lines_sorted(output_file)
+    self.assertListEqual(expected, actual)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py:100]: do we need to do any rounding in `report_predictions` to avoid flakes due to possible lack of precision?

**Comment by ryanthompson591** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+from apache_beam.ml.inference.base import RunInference
+from apache_beam.ml.inference.sklearn_inference import ModelFileType
+from apache_beam.ml.inference.sklearn_inference import SklearnModelHandlerPandas
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+import pandas
+
+MODELS = [{
+    'name': 'all_features',
+    'required_features': [
```
[the remark this answers is not in the document]
Comment (ryanthompson591) [reply sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py]: Good suggestion. This was bothering me when yapf would fix this. I was able to fix this with # yapf: disable

**Comment by ryanthompson591** [comment]
Run Python 3.7 PostCommit

**Comment by ryanthompson591** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py (hunk 11 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/sklearn_inference_it_test.py]: not sure, will trust your judgement here.

**Comment by tvalentyn** [comment]
Let me know when this is ready to review/merge.

**Comment by ryanthompson591** [comment]
Run Python 3.9 PostCommit

**Comment by ryanthompson591** [comment]
Run Python 3.9 PostCommit

**Comment by ryanthompson591** [comment]
Run Python 3.7 PostCommit

**Comment by ryanthompson591** [comment]
Run Python 3.9 PostCommit

File: sdks/python/apache_beam/examples/inference/README.md [hunk 13]
```diff
@@ -255,4 +255,42 @@ This writes the output to the `predictions.txt` with contents like:
...
+```sh
+python -m apache_beam.examples.inference.sklearn_japanese_housing_regression.py \
+  --input_file INPUT \
+  --output OUTPUT \
+  --model_path MODEL_PATH
+```
+For example:
+```sh
+python -m apache_beam.examples.inference.sklearn_japanese_housing_regression.py \
+  --input_file mnist_data.csv \
```
[the remark this answers is not in the document]
Comment (ryanthompson591) [reply sdks/python/apache_beam/examples/inference/README.md:298]: Let me give a better name.

**Comment by ryanthompson591** [comment]
@tvalentyn This looks ready to merge, the failing test is not the new test I added but rather the issue described here: https://github.com/apache/beam/issues/22605

**Comment by ryanthompson591** [comment]
Run Python 3.9 PostCommit

File: sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py [hunk 14]
```diff
@@ -32,50 +32,55 @@
...
 
 def sort_by_features(dataframe, max_size):
-  """ Partitions the dataframe by what data it has available."""
+  """ Returns an index to a model, based on available data."""
   for i, model in enumerate(MODELS):
     required_features = dataframe[model['required_features']]
-    if not required_features.isnull().any().any():
+    # A model can only make a prediction if all required features
+    # are present.
+    # required_features is 2D single row, so all must be called twice.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/sklearn_japanese_housing_regression.py:82]:
```suggestion
    # required_features is 2D single row, so all() must be called twice.
```

File: sdks/python/apache_beam/ml/inference/sklearn_inference_test.py [hunk 15]
```diff
@@ -47,6 +47,12 @@
 from apache_beam.testing.util import assert_that
 from apache_beam.testing.util import equal_to
 
+# pylint: disable=wrong-import-order, wrong-import-position, ungrouped-imports, unused-import
+try:
+  from apache_beam.io.gcp.gcsfilesystem import GCSFileSystem
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/sklearn_inference_test.py:52]: why is this necessary? I am not seeing any usage of GCSFileSystem

**Review by tvalentyn — APPROVED** [review]

File: sdks/python/apache_beam/ml/inference/sklearn_inference_test.py (hunk 15 above)
Comment (ryanthompson591) [reply sdks/python/apache_beam/ml/inference/sklearn_inference_test.py]: This might have been preemptive. Unless we are able to save files, I think these tests could fail.  However, it doesn't seem relevant to this PR and we should defer fixing it until it becomes an issue.

**Review by ryanthompson591 — COMMENTED** [review]
@tvalentyn PTAL

**Comment by ryanthompson591** [comment]
Run Python 3.9 PostCommit

**Comment by tvalentyn** [comment]
Run Python 3.9 PostCommit
